### ISSUE TITLE
Ability to set backbuffer size

### ISSUE THREAD

**Issue by fatcerberus** [liballeg/allegro5#779]
When using `ALLEGRO_FULLSCREEN_WINDOW`, the user has no control over the display's resolution, as it's dependent on the OS setting.  For games with a fixed nominal resolution, this requires using a scaling transformation to get everything to match in fullscreen.  However, there are a few issues:

* Resolution-matching by scaling breaks some primitives such as `ALLEGRO_PRIM_LINE_LIST`, which get rendered at a higher resolution than the rest of the game and thus look weird.  This forces lines and points to be drawn as quads instead, which is more complicated.
* Scaling by a non-integer factor causes interpolation artifacts when drawing only parts of a bitmap, as the GPU will sample texels outside of the designated area.  This is a big problem for games using texture atlases.

So here's my feature request: What I propose is for Allegro to (perhaps optionally) provide a custom backbuffer whose size can be set independently of the window size, so games can do pixel-perfect rendering regardless of the actual size of the display.

Since Allegro bitmaps are FBOs, this can of course be done by the user, by doing all rendering to a bitmap of the desired size and then drawing that to the real backbuffer just before the flip.  But it would be great if Allegro could do this automatically; it'd be particularly useful for developers of retro games.

**Comment by SiegeLord** [comment]
A few comments on this:
- This needs to be independent of FULLSCREEN_WINDOW, as you probably want the same feature to work with windowed and even regular fullscreen displays.
- When the real backbuffer and the proposed fake backbuffer differ in size, there's a question of how to convert from one to the other. Do you scale one to fit the other? Do letter boxing? If letter boxing, what to do about the unfilled regions?
   - Granted, we can just choose an option and leave the user to do something else if they want to configure it
- Most critically, what about all the screen coordinate APIs? E.g. mouse coordinates?

Anyway, despite being sympathetic to the feature, it just feels odd to place it inside Allegro's core. It feels like it could belong in a utility library, maybe an addon. There have been a few such utility function ideas (e.g. a al_init_everything, or al_run_main_loop) that could find their place here.

Lastly, here's a possible API for this. Here I am going for an implementation that does not touch Allegro's internals (although, it makes it less transparent than it could in principle be):

```c
ALLEGRO_DISPLAY* d = al_create_display(...);
// Also sets the target bitmap for convenience?
VIEWPORT* v = create_viewport(al_get_backbuffer(d), new_width, new_height, LETTERBOX);

// ...
if (event.type == ALLEGRO_EVENT_MOUSE_AXES) {
  int mx, my;
  transform_coordinates_to_viewport(event.mouse.x, event.mouse.y, &mx, &my);
}

flip_viewport(v);
// Alternatively
draw_viewport(v);
al_flip_display();
```

**Comment by fatcerberus** [comment]
> This needs to be independent of FULLSCREEN_WINDOW, as you probably want the same feature to work with windowed and even regular fullscreen displays.

I agree with this, it may be useful when making retro games to want to render at some low resolution (like 320x240) and then scale up by 2x or 3x to show the pixels.

Regarding scaling: The most useful default, IMO, would be to scale up to fill the display but maintain the aspect ratio by letterboxing.  If the scaling mode was not customizable, that's definitely the method I would pick.

> Most critically, what about all the screen coordinate APIs? E.g. mouse coordinates?

Thanks for bringing this up; I recently implemented something like the feature proposed here in miniSphere and I'm pretty sure my mouse API is now buggy. :wink:  And I was very close to making a release too!

**Comment by fatcerberus** [comment]
Seems like this issue is a duplicate of #336, I wasn't the first person to think of it :)

**Comment by fatcerberus** [comment]
> Anyway, despite being sympathetic to the feature, it just feels odd to place it inside Allegro's core. It feels like it could belong in a utility library, maybe an addon. There have been a few such utility function ideas (e.g. a al_init_everything, or al_run_main_loop) that could find their place here.

I see what you mean; however, the reason I suggested it as a core feature is that more and more platforms can only support `ALLEGRO_FULLSCREEN_WINDOW` which doesn't allow the game to control the actual resolution.  As a library designed used for 2D raster drawing, pixel-perfect should always be the default--and that's not possible with ALLEGRO_FULLSCREEN_WINDOW *unless* you use a fake backbuffer.

**Comment by elias-pschernig** [comment]
I think the default mode in Allegro would not be an extra buffer but using the transformation API. Just like I thought in 2010 apparently :) The only exception would be retro-style low-resolution pixel games. Anyway, just like what was suggest back then we could make it an option - and I think a VIEWPORT class like SiegeLord suggests makes most sense. You could then even do things like have 2 or 4 viewports for split-screen games, or render to off-screen viewports.

**Comment by fatcerberus** [comment]
I used transforms to do scaling in miniSphere up until a few releases ago and there were two problems:
* Lines and points drawn with al_draw_prim are at the native screen resolution making them look weird amongst scaled-up bitmaps
* (Bigger issue) Texture atlases don't work: When drawing only part of a bitmap using a scale transform, scaling by a non-integer factor causes the GPU to blend texels which will be very surprising to users if it's the default behavior.

In light of the second point (scaling the screen affects behavior for offscreen surfaces in a non-obvious way), I definitely wouldn't recommend transform-based scaling by default.

**Comment by elias-pschernig** [comment]
Yes, for pixel art you definitely want a buffer!

**Comment by fatcerberus** [comment]
Not only pixel art but just multi-image textures in general, e.g. tilesets.  If you scale you get pixels from adjacent tiles which looks rather ugly.

**Comment by elias-pschernig** [comment]
Well, there's standard ways to deal with that. Depending on the type of filtering you need a border of 1, 2 or more pixels around each texture. And depending on the type of texture you want to repeat the border pixels, the opposite border pixels, or transparency inside of that border. It's just in the nature of things and independent of Allegro, you would do the same thing with DirectX, OpenGL and probably even Unity.

**Comment by fatcerberus** [comment]
> It's just in the nature of things and independent of Allegro, you would do the same thing with DirectX, OpenGL and probably even Unity.

Oh definitely; I'm just saying that if there were hypothetically going to be a default scaling behavior for cases where the nominal resolution doesn't match the actual screen resolution (because the mismatch is unavoidable with fixed resolutions combined with FULLSCREEN_WINDOW), I would consider the two caveats listed above to violate the principle of least astonishment.  A pixel-perfect buffer would be much more predictable in that regard.

...then again, my *own* use case is far from typical for Allegro, being not a standalone game but rather a general-purpose JavaScript-powered game *engine*, so what do I know? 😄

**Comment by beoran** [comment]
I would not add such a feature to the core of Allegro, but rather to an
official add-on with other ease of use utility functions and structures
like Siegelord mentioned. It seems to me that some people are still
intimidated by the Allegro5 API, and an official easier wrapper add-on
would be helpful.

**Comment by fatcerberus** [comment]
See, I would 100% agree with not adding it to the core, if it weren't for the fact that many platforms now don't support native fullscreen at all, forcing you to use FULLSCREEN_WINDOW and thereby lose control over the size of your backbuffer.

**Comment by beoran** [comment]
AFAIK the backbuffer always has the same size as the current display mode.
While in theory we could probably provide a function like ALLEGRO_BITMAP *
al_create_backbuffer(ALLEGRO_DISPLAY * d, int w, int h), I am not quite
sure how that would work out in practice.

But if you or someone else makes a PR with such a feature and some
examples, I would likely be in favour of it.
